File: I-160-2/Quilting.cs
using System;
using System.Collections;
using System.Globalization;
using System.Text;
using System.Text.RegularExpressions;

public class Quilting 
{
	int probe( int x, int y, int[,] M, int length, int width )
	{
		if ( x < 0 || y < 0 || x >= width || y >= length ) return -2;
		return M[x, y];
	}
	public string lastPatch( int length, int width, string[] colorList ) 
	{
		int x = width / 2;
		int y = length / 2;
		int[] dx = new int[] {0, -1, 0, 1};
		int[] dy = new int[] {-1, 0, 1, 0};
		int[,] M = new int[width, length];
		for( int i = 0; i < width; i++ )
			for( int j = 0; j < length; j++ )
				M[i, j] = -1;
		M[x, y] = 0;
		int dir = 0;
		int step = 0;
		int[] c = new int[colorList.Length];
		c[0]++;
		for( int k = 1; k < length * width; )
		{
			if ( dir % 2 == 0 )
				step++;
			for( int l = 0; k < length * width && l < step; l++ )
			{
				k++;
				x += dx[dir];
				y += dy[dir];
				int[] colors = new int[colorList.Length];
				for( int i = x - 1; i <= x + 1; i++ )
					for( int j = y - 1; j <= y + 1; j++ )
						if ( probe( i, j, M, length, width ) > -1 )
							colors[M[i, j]]++;
				int min = colors[0] * length * width + c[0];
				int newcolor = 0;
				for( int i = 1; i < colorList.Length; i++ )
					if ( min > colors[i] * length * width + c[i] )
					{
						min = colors[i] * length * width + c[i];
						newcolor = i;
					}
				c[newcolor]++;
				M[x, y] = newcolor;
			}
			dir = (dir + 1) % 4;
		}
		return colorList[M[x, y]];
	}

	NumberFormatInfo nfi = new CultureInfo( "en-US", false ).NumberFormat;

// BEGIN CUT HERE
	public static void Main(string[] args)
	{
		DateTime t1 = DateTime.Now;
		eq(1,(new Quilting()).lastPatch(3, 2, new String[] {"RED","BLUE","TAN"}),"TAN");
		eq(2,(new Quilting()).lastPatch(4, 3, new String[] {"E","D","C","B","A"}),"E");
		eq(3,(new Quilting()).lastPatch(3, 3, new String[] {"A","B","C","D"}),"C");
		eq(4,(new Quilting()).lastPatch(1, 1, new String[] {"RED","BLUE","YELLOW"}),"RED");
		eq(5,(new Quilting()).lastPatch(10, 10, new String[] {"X","Y","Z"}),"Z");
		DateTime t2 = DateTime.Now;
		TimeSpan t = t2 - t1;
		Console.WriteLine( "{0} sec", t.TotalMilliseconds / 1000 ); 
	}
	private static void eq(int n, int a, int b)
	{
		if(a == b)
			Console.WriteLine("Case "+n+" passed.");
		else
			Console.WriteLine("Case "+n+" failed: expected \""+b+"\", received \""+a+"\".");
	}
	private static void eq(int n, double a, double b)
	{
		if(a == b)
			Console.WriteLine("Case "+n+" passed.");
		else
			Console.WriteLine("Case "+n+" failed: expected \""+b+"\", received \""+a+"\".");
	}
	private static void eq(int n, char a, char b)
	{
		if(a == b)
			Console.WriteLine("Case "+n+" passed.");
		else
			Console.WriteLine("Case "+n+" failed: expected '"+b+"', received '"+a+"'.");
	}
	private static void eq(int n, long a, long b)
	{
		if(a == b)
			Console.WriteLine("Case "+n+" passed.");
		else
			Console.WriteLine("Case "+n+" failed: expected \""+b+"\", received \""+a+"\".");
	}
	private static void eq(int n, bool a, bool b)
	{
		if(a == b)
			Console.WriteLine("Case "+n+" passed.");
		else
			Console.WriteLine("Case "+n+" failed: expected \""+b+"\", received \""+a+"\".");
	}
	private static void eq(int n, string a, string b)
	{
		if(a == b)
			Console.WriteLine("Case "+n+" passed.");
		else
			Console.WriteLine("Case "+n+" failed: expected \""+b+"\", received \""+a+"\".");
	}
	private static void eq(int n, int[] a, int[] b)
	{
		if(a.Length != b.Length)
		{
			Console.WriteLine("Case "+n+" failed: returned "+a.Length+" elements; expected "+b.Length+" elements.");
			return;
		}
		for(int i= 0; i < a.Length; i++)
			if(a[i] != b[i])
			{
				Console.WriteLine("Case "+n+" failed. Expected and returned array differ in position "+i);
				print(b);
				print(a);
				return;
			}
		Console.WriteLine("Case "+n+" passed.");
	}
	private static void eq(int n, double[] a, double[] b)
	{
		if(a.Length != b.Length)
		{
			Console.WriteLine("Case "+n+" failed: returned "+a.Length+" elements; expected "+b.Length+" elements.");
			return;
		}
		for(int i= 0; i < a.Length; i++)
			if(a[i] != b[i])
			{
				Console.WriteLine("Case "+n+" failed. Expected and returned array differ in position "+i);
				print(b);
				print(a);
				return;
			}
		Console.WriteLine("Case "+n+" passed.");
	}
	private static void eq(int n, long[] a, long[] b)
	{
		if(a.Length != b.Length)
		{
			Console.WriteLine("Case "+n+" failed: returned "+a.Length+" elements; expected "+b.Length+" elements.");
			return;
		}
		for(int i= 0; i < a.Length; i++)
			if(a[i] != b[i])
			{
				Console.WriteLine("Case "+n+" failed. Expected and returned array differ in position "+i);
				print(b);
				print(a);
				return;
			}
		Console.WriteLine("Case "+n+" passed.");
	}
	private static void eq(int n, string[] a, string[] b)
	{
		if(a.Length != b.Length)
		{
			Console.WriteLine("Case "+n+" failed: returned "+a.Length+" elements; expected "+b.Length+" elements.");
			return;
		}
		for(int i= 0; i < a.Length; i++)
			if(a[i] != b[i])
			{
				Console.WriteLine("Case "+n+" failed. Expected and returned array differ in position "+i);
				print(b);
				print(a);
				return;
			}
		Console.WriteLine("Case "+n+" passed.");
	}
	private static void print(int a)
	{
		Console.Write(a+" ");
	}
	private static void print(string s)
	{
		Console.Write("\""+s+"\" ");
	}
	private static void print(int[] rs)
	{
		if(rs == null) return;
		Console.Write('{');
		for(int i= 0; i < rs.Length; i++)
		{
			Console.Write(rs[i]);
			if(i != rs.Length-1)
				Console.Write(", ");
		}
		Console.WriteLine('}');
	}
	private static void print(double[] rs)
	{
		if(rs == null) return;
		Console.Write('{');
		for(int i= 0; i < rs.Length; i++)
		{
			Console.Write(rs[i]);
			if(i != rs.Length-1)
				Console.Write(", ");
		}
		Console.WriteLine('}');
	}
	private static void print(long[] rs)
	{
		if(rs == null) return;
		Console.Write('{');
		for(int i= 0; i < rs.Length; i++)
		{
			Console.Write(rs[i]);
			if(i != rs.Length-1)
				Console.Write(", ");
		}
		Console.WriteLine('}');
	}
	private static void print(string[] rs)
	{
		if(rs == null) return;
		Console.Write('{');
		for(int i= 0; i < rs.Length; i++)
		{
			Console.Write("\""+rs[i]+"\"");
			if(i != rs.Length-1)
				Console.Write(", ");
		}
		Console.WriteLine('}');
	}
	private static void nl()
	{
		Console.WriteLine();
	}
// END CUT HERE
}

File: II-160-3/Quilting.cpp
#include <algorithm>
#include <cmath>
#include <cstdlib>
#include <functional>
#include <iostream>
#include <iterator>
#include <map>
#include <numeric>
#include <set>
#include <sstream>
#include <stack>
#include <string>
#include <queue>
#include <vector>
using namespace std;
// BEGIN CUT HERE
#include <Windows.h>
#pragma comment( lib, "Winmm" )
// END CUT HERE
typedef vector<int> vi;
typedef vector<vi> vvi;
typedef vector<string> vs;
#define fora(i,v) for(int i = 0; i < (v).size(); i++)
#define all(v) (v).begin(), (v).end()
struct Quilting
{
	string lastPatch( int length, int width, vs colorList )
	{
		int dx[] = {-1, 0, 1, 0};
		int dy[] = {0, -1, 0, 1};
		int n = colorList.size(), x = length / 2, y = width / 2, d = 0, step = 0;
		vi colors(n, 0); ++colors[0];
		vvi M(length, vi(width, -1)); M[x][y] = 0;
		for( int i = 1; i < width * length; )
		{
			if ( d % 2 == 0 )
				++step;
			for( int j = 0; j < step && i < width * length; ++j )
			{
				++i;
				x += dx[d]; y += dy[d];
				vi neighbors(n, 0);
				for( int a = x - 1; a <= x + 1; ++a )
					if ( a >= 0 && a < length )
						for( int b = y - 1; b <= y + 1; ++b )
							if ( b >= 0 && b < width && M[a][b] >= 0 )
								++neighbors[M[a][b]];
				int k = *min_element( all(neighbors) );
				int c = -1;
				fora(z, neighbors)
					if ( neighbors[z] == k && (c == -1 || colors[c] > colors[z]) )
						c = z;
				M[x][y] = c; ++colors[c];
			}
			d = (d + 1) % 4;
		}
		return colorList[M[x][y]];
	}
	// BEGIN CUT HERE
	private:
	template <typename T> string print_array( const vector<T>& v ) { ostringstream os; os << "{ "; for( typename vector<T>::const_iterator it = v.begin(); it != v.end(); ++it ) os << '\"' << *it << "\","; os << " }"; return os.str(); }
	template<typename T> void verify_case( int Case, const vector<T>& expected, const vector<T>& received ) { cerr << "Test Case #" << Case << "..."; if ( expected == received ) cerr << "PASSED. "; else { cerr << "FAILED." << endl; cerr << "\tExpected: " << print_array( expected ) << endl; cerr << "\tReceived: " << print_array( received ) << endl; } }
	void verify_case( int Case, const long long& expected, const long long& received ) { cerr << "Test Case #" << Case << "..."; if ( expected == received ) cerr << "PASSED. "; else { cerr << "FAILED." << endl; cerr << "\tExpected: \"" << expected << '\"' << endl; cerr << "\tReceived: \"" << received << '\"' << endl; } }
	void verify_case( int Case, const double& expected, const double& received ) { cerr << "Test Case #" << Case << "..."; if ( expected == received ) cerr << "PASSED. "; else { cerr << "FAILED." << endl; cerr << "\tExpected: \"" << expected << '\"' << endl; cerr << "\tReceived: \"" << received << '\"' << endl; } }
	void verify_case( int Case, const int& expected, const int& received ) { cerr << "Test Case #" << Case << "..."; if ( expected == received ) cerr << "PASSED. "; else { cerr << "FAILED." << endl; cerr << "\tExpected: \"" << expected << '\"' << endl; cerr << "\tReceived: \"" << received << '\"' << endl; } }
	void verify_case( int Case, const string& expected, const string& received ) { cerr << "Test Case #" << Case << "..."; if ( expected == received ) cerr << "PASSED. "; else { cerr << "FAILED." << endl; cerr << "\tExpected: \"" << expected << '\"' << endl; cerr << "\tReceived: \"" << received << '\"' << endl; } }
	public:
	void run_test( int Case = -1 ) { vector<int> A0; vector<int> A1; vector<vs> A2; vector<string> A3;
	{ /*Test 0*/int length = 
	3
	; A0.push_back( length ); int width = 
	2
	; A1.push_back( width ); string _colorList[] = 
	{"RED","BLUE","TAN"}
	; vs colorList(_colorList, _colorList + (sizeof(_colorList) / sizeof(_colorList[0]))); A2.push_back( colorList ); string result = 
	"TAN"
	; A3.push_back( result );  }
	{ /*Test 1*/int length = 
	4
	; A0.push_back( length ); int width = 
	3
	; A1.push_back( width ); string _colorList[] = 
	{"E","D","C","B","A"}
	; vs colorList(_colorList, _colorList + (sizeof(_colorList) / sizeof(_colorList[0]))); A2.push_back( colorList ); string result = 
	"E"
	; A3.push_back( result );  }
	{ /*Test 2*/int length = 
	3
	; A0.push_back( length ); int width = 
	3
	; A1.push_back( width ); string _colorList[] = 
	{"A","B","C","D"}
	; vs colorList(_colorList, _colorList + (sizeof(_colorList) / sizeof(_colorList[0]))); A2.push_back( colorList ); string result = 
	"C"
	; A3.push_back( result );  }
	{ /*Test 3*/int length = 
	1
	; A0.push_back( length ); int width = 
	1
	; A1.push_back( width ); string _colorList[] = 
	{"RED","BLUE","YELLOW"}
	; vs colorList(_colorList, _colorList + (sizeof(_colorList) / sizeof(_colorList[0]))); A2.push_back( colorList ); string result = 
	"RED"
	; A3.push_back( result );  }
	{ /*Test 4*/int length = 
	10
	; A0.push_back( length ); int width = 
	10
	; A1.push_back( width ); string _colorList[] = 
	{"X","Y","Z"}
	; vs colorList(_colorList, _colorList + (sizeof(_colorList) / sizeof(_colorList[0]))); A2.push_back( colorList ); string result = 
	"Z"
	; A3.push_back( result );  }
	for( int i = 0; i < A0.size(); i++ ) if ( Case == -1 || i == Case ) { DWORD start, finish; start = timeGetTime(); verify_case( i, A3[i], lastPatch( A0[i], A1[i], A2[i]) ); finish = timeGetTime(); cout << "Elapsed " << finish - start << " msec" << endl; } }
// END CUT HERE
};
// BEGIN CUT HERE
int main()
{
	DWORD start, finish;
	start = timeGetTime();
	Quilting ___test; 
	___test.run_test(); 
	finish = timeGetTime();
	cout << endl << "Elapsed " << finish - start << " msec" << endl;
} 
// END CUT HERE
